hugr-model/src/v0/scope/symbol.rs

```diff
@@ -0,0 +1,208 @@
+            if shadowed_binding.scope_depth == scope_depth {
+                self.symbols.insert(name, shadowed);
+                return Err(DuplicateSymbolError(name.into(), node, *shadowed_node));
+            }
+        }
+
+        self.bindings.insert(
+            node,
+            Binding {
+                scope_depth,
+                shadows: shadowed,
+                symbol_index,
+            },
+        );
+
+        Ok(())
+    }
+
+    /// Check whether a symbol is currently visible in the current scope.
+    pub fn is_visible(&self, node: NodeId) -> bool {
+        let Some(binding) = self.bindings.get(&node) else {
+            return false;
+        };
+
+        // Check that the symbol has not been shadowed at this point.
+        self.symbols[binding.symbol_index] == binding.symbol_index
+    }
+
+    /// Tries to resolve a symbol name in the current scope.
+    pub fn resolve(&self, name: &'a str) -> Result<NodeId, UnknownSymbolError> {
+        let index = *self
+            .symbols
+            .get(name)
+            .ok_or(UnknownSymbolError(name.into()))?;
+
+        // NOTE: The unwrap is safe because the `symbols` map
+        // points to valid indices in the `bindings` map.
+        let (node, _) = self.bindings.get_index(index).unwrap();
+        Ok(*node)
+    }
+
+    /// Returns the depth of the given region, if it corresponds to a currently open scope.
+    pub fn region_to_depth(&self, region: RegionId) -> Option<ScopeDepth> {
+        Some(self.scopes.get_index_of(&region)? as _)
+    }
+
+    /// Returns the region corresponding to the scope at the given depth.
+    pub fn depth_to_region(&self, depth: ScopeDepth) -> Option<RegionId> {
+        let (region, _) = self.scopes.get_index(depth as _)?;
+        Some(*region)
+    }
+
+    /// Resets the symbol table to its initial state while maintaining its
+    /// allocated memory.
+    pub fn clear(&mut self) {
+        self.symbols.clear();
+        self.bindings.clear();
+        self.scopes.clear();
+    }
+}
+
+impl Default for SymbolTable<'_> {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+#[derive(Debug, Clone, Copy)]
+struct Binding {
+    /// The depth of the scope in which this binding is defined.
+    scope_depth: ScopeDepth,
+
+    /// The index of the binding that is shadowed by this one, if any.
+    shadows: Option<BindingIndex>,
+
+    /// The index of this binding's symbol in the symbol table.
+    ///
+    /// The symbol table always points to the currently visible binding for a
+    /// symbol. Therefore this index is only valid if this binding is not shadowed.
+    /// In particular, we detect shadowing by checking if the entry in the symbol
+    /// table at this index does indeed point to this binding.
+    symbol_index: SymbolIndex,
+}
+
+#[derive(Debug, Clone, Copy)]
+struct Scope {
+    /// The length of the `bindings` stack when this scope was entered.
```

Comment:
TBH I think this comment is the key here - you could just put this on `SymbolTable::scopes` and then elide `struct Scope` altogether...